library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java

```diff
@@ -26,183 +27,299 @@
+    }
+    //</editor-fold>
+
+    //<editor-fold desc="seperate onFailedToRecycleView">
+    @Override
+    public final boolean onFailedToRecycleView(VH holder) {
+        int position = holder.getAdapterPosition();
+        boolean ret;
+
+        if (isHeaderPosition(position)) {
+            ret = onFailedToRecycleHeaderView(holder);
+        } else if (isFooterPosition(position)) {
+            ret = onFailedToRecycleFooterView(holder);
+        } else {
+            ret = onFailedToRecycleItemView(holder);
+        }
+
+        return ret;
+    }
+
+    protected boolean onFailedToRecycleHeaderView(VH holder) {
+        return false;
+    }
+
+    protected boolean onFailedToRecycleItemView(VH holder) {
+        return false;
+    }
+
+    protected boolean onFailedToRecycleFooterView(VH holder) {
+        return false;
+    }
+    //</editor-fold>
+
+    //<editor-fold desc="seperate onViewAttachedToWindow">
+    @Override
+    public final void onViewAttachedToWindow(VH holder) {
+        int position = holder.getAdapterPosition();
+
+        if (isHeaderPosition(position)) {
+            onHeaderViewAttachedToWindow(holder);
+        } else if (isFooterPosition(position)) {
+            onFooterViewAttachedToWindow(holder);
+        } else {
+            onItemViewAttachedToWindow(holder);
+        }
+    }
+
+    protected void onHeaderViewAttachedToWindow(VH holder) {
+    }
+
+    protected void onItemViewAttachedToWindow(VH holder) {
+    }
+
+    protected void onFooterViewAttachedToWindow(VH holder) {
+    }
+    //</editor-fold>
+
+    //<editor-fold desc="seperate onViewDetachedFromWindow">
+    @Override
+    public final void onViewDetachedFromWindow(VH holder) {
+        int position = holder.getAdapterPosition();
+
+        if (isHeaderPosition(position)) {
+            onHeaderViewDetachedFromWindow(holder);
+        } else if (isFooterPosition(position)) {
+            onFooterViewDetachedFromWindow(holder);
+        } else {
+            onItemViewDetachedFromWindow(holder);
+        }
+    }
+
+    protected void onHeaderViewDetachedFromWindow(VH holder) {
+    }
+
+    protected void onItemViewDetachedFromWindow(VH holder) {
+    }
+
+    protected void onFooterViewDetachedFromWindow(VH holder) {
+    }
+    //</editor-fold>
+
+    //<editor-fold desc="seperate onViewRecycled">
+    @Override
+    public final void onViewRecycled(VH holder) {
+        int position = holder.getAdapterPosition();
+
+        if (isHeaderPosition(position)) {
+            onHeaderViewRecycled(holder);
+        } else if (isFooterPosition(position)) {
+            onFooterViewRecycled(holder);
+        } else {
+            onItemViewRecycled(holder);
+        }
+    }
+
+    protected void onHeaderViewRecycled(VH holder) {
+    }
+
+    protected void onItemViewRecycled(VH holder) {
+    }
+
+    protected void onFooterViewRecycled(VH holder) {
+    }
+    //</editor-fold>
```

Comment:
To keep the library simplicity we can also remove these methods related to: ``onViewDetachedFromWindow`` and ``onViewRecycled``